### ISSUE TITLE
[Feature]: MXFP8 SDPA backward performance and efficient variable-length support

### ISSUE THREAD

**Issue by ZejiaZheng** [NVIDIA/cudnn-frontend#659]
### Before submitting

- [x] I searched existing issues and did not find a duplicate.

### Feature request
<h2><span class="s1"><b>Summary</b></span></h2>

<p class="p3">We are evaluating cuDNN MXFP8 SDPA for training on GB300 and are seeing very good performance from the forward path, but two issues currently limit its usefulness for training workloads:</p>

<ol start="1"><li><span class="s1"><b>MXFP8 backward provides little speedup over BF16</b></span>, despite the substantial forward speedup.</li><li><span class="s1"><b>Enabling </b><b><code>use_padding_mask</code></b><b> / </b><b><code>seq_len_*</code></b><b> adds ~35% forward overhead even when no tokens are actually masked.</b></span></li></ol>

<p class="p4">For our workload, these two issues eliminate much of the end-to-end benefit of using MXFP8 attention.</p>

<p class="p3">We would like to request:</p>

<ul><li>Improved MXFP8 SDPA backward performance.</li><li>Efficient support for variable-length / packed MXFP8 attention, ideally directly through <code>sdpa_mxfp8</code>.</li><li>Reduced overhead for the <code>sdpa_fp8</code> padding-mask path when <code>seq_len_q == seq_len_kv == s</code>.</li></ul>

<p class="p4">A standalone repro is attached: <code>mxfp8_sdpa_repro.py</code>.</p>

<h2><span class="s1"><b>Workload</b></span></h2>

<p class="p5">Our representative attention geometry is:</p>

<pre><code class="language-text">GPU:       NVIDIA GB300 (SM103)
Q heads:   88
KV heads:  8
GQA:       11:1
head_dim:  128
batch:     1
causal:    false
attention: full self-attention</code></pre>

<p class="p1">This comes from a large MoE video-diffusion training workload.</p>

<h2><span class="s1"><b>1. MXFP8 backward performance</b></span></h2>

<p class="p4">MXFP8 forward performs very well on GB300:</p>

Seq len | MXFP8 fwd | BF16 fwd | Speedup
-- | -- | -- | --
2,048 | 0.087 ms / 2171 TF/s | 0.139 ms / 1360 TF/s | 1.60x
4,096 | 0.302 ms / 2500 TF/s | 0.437 ms / 1728 TF/s | 1.45x
8,192 | 1.094 ms / 2764 TF/s | 1.509 ms / 2003 TF/s | 1.38x
16,384 | 4.369 ms / 2768 TF/s | 6.297 ms / 1921 TF/s | 1.44x

<p class="p1">The forward reaches roughly <span class="s1"><b>2.2–2.8 PFLOP/s</b></span>, while backward reaches only <span class="s1"><b>1.0–1.6 PFLOP/s</b></span>.</p>

<p class="p3">In other words:</p>

<pre><code class="language-text">MXFP8 forward:   ~1.38–1.60x BF16
MXFP8 backward:  ~1.00–1.11x BF16</code></pre>

<p class="p1">This makes MXFP8 considerably less attractive for training than the forward numbers alone would suggest.</p>

<h3><span class="s1"><b>Request</b></span></h3>

<p class="p1">Could the MXFP8 backward kernels be optimized to obtain a benefit closer to the forward path?</p>

<p class="p1">We would also appreciate guidance on whether the current backward throughput is expected for GB300/SM103, or whether this indicates a missing optimization / kernel selection issue.</p>

<h2><span class="s1"><b>2. Padding-mask / variable-length overhead</b></span></h2>

<p class="p1">Our training workload uses packed variable-length sequences.</p>

<p class="p4">For MXFP8, <code>graph.sdpa_mxfp8</code> currently does not expose:</p>

<pre><code class="language-python">use_padding_mask
seq_len_q
seq_len_kv</code></pre>

<p class="p1">We can reach the same MXFP8 path through <code>graph.sdpa_fp8</code> using E8M0 / <code>F8_128x4</code> descales and enable the padding mask there.</p>

<p class="p1">The masking itself appears to work correctly: reducing the effective sequence length by 2 reduces attention work by approximately 4x.</p>

<p class="p1">However, simply enabling the padding-mask path introduces substantial overhead even when nothing is masked.</p>

<p class="p3">For:</p>

<pre><code class="language-text">b=4
s=8192
h_q=88
h_kv=8
d=128</code></pre>

<p class="p1">we measure:</p>

<pre><code class="language-text">sdpa_mxfp8, no mask                     4.33 ms

sdpa_fp8 + padding mask
  seq_len = s                           5.87 ms   (+35%)

sdpa_fp8 + padding mask
  seq_len = s/2                         1.63 ms</code></pre>

<p class="p1">The <code>seq_len = s/2</code> result is encouraging: fully masked tiles appear to be skipped efficiently.</p>

<p class="p3">The problematic case is:</p>

<pre><code class="language-text">seq_len = s</code></pre>

<p class="p1">where the computational workload should effectively be identical to the unmasked case, but enabling the padding-mask path costs approximately <span class="s1"><b>35%</b></span>.</p>

<p class="p1">This overhead consumes most of the MXFP8 forward speedup relative to BF16 for packed/variable-length training.</p>

<h3><span class="s1"><b>Request</b></span></h3>

<p class="p1">We would like an efficient MXFP8 variable-length path with minimal overhead when sequences are close to the maximum length.</p>

<p class="p4">Ideally:</p>

<pre><code class="language-python">graph.sdpa_mxfp8(
    ...,
    use_padding_mask=True,
    seq_len_q=...,
    seq_len_kv=...,
)</code></pre>

<p class="p1">would be supported directly and would retain performance close to the unmasked MXFP8 kernel when little or no padding is present.</p>

<p class="p1">Alternatively, optimizing the existing <code>sdpa_fp8</code> padding-mask path would address the same issue.</p>

<h2><span class="s1"><b>Expected impact</b></span></h2>

<p class="p1">For training, the relevant cost is forward + backward rather than forward alone.</p>

<p class="p1">MXFP8 forward already demonstrates that the hardware and kernel can provide a substantial improvement over BF16. Improving backward throughput and removing the padding-mask overhead would make MXFP8 SDPA significantly more useful for large-scale variable-length training workloads.</p>

<h2><span class="s1"><b>Correctness</b></span></h2>

<p class="p4">The repro uses the same MXFP8 quantization/scaling conventions as the cuDNN frontend MXFP8 SDPA tests:</p>

<ul><li>E4M3 data</li><li>E8M0 block scaling</li><li>32-element blocks</li><li><code>F8_128x4</code><span class="s2"> scale swizzle</span></li></ul>

<p class="p1">Forward output was also checked against a dequantized BF16 reference.</p>

<p class="p5">Observed relative mean error is approximately:</p>

<pre><code class="language-text">2.7%</code></pre>

<p class="p1">which is consistent with the expected quantization error.</p>

<h2><span class="s1"><b>Environment</b></span></h2>

<pre><code class="language-text">GPU              NVIDIA GB300
compute capability 10.3
GPU memory        277 GiB
max SM clock      2070 MHz
power limit       1400 W
GPUs/node         4

driver            580.159.03
CUDA              13.0
nvcc              13.0.88

host              Ubuntu 24.04.3 LTS
kernel            Linux 6.14.0-1008-nvidia-64k
architecture      aarch64
CPU               ARM Neoverse-V2, 144 cores

Python            3.12.12
PyTorch           2.11.0+cu130
cuDNN             9.24.0
nvidia-cudnn-cu13 9.24.0.43
cudnn-frontend    1.26.0
                  1.27.0 tested with identical results

nvidia-cutlass-dsl 4.6.0.dev0
triton             3.6.0
torchao            0.17.0
flash-attn-4        4.0.0b23.dev2+g2409214</code></pre>

<p class="p1">The BF16 numbers above use FlashAttention 4 / CuTe DSL on SM100+.</p>

<p class="p1">The repro can also use PyTorch SDPA so that FlashAttention is not required.</p>

<h2><span class="s1"><b>Reproduction</b></span></h2>

<p class="p4">Minimal installation:</p>

<pre><code class="language-bash">pip install torch nvidia-cudnn-frontend==1.26.0</code></pre>

<p class="p1">Basic run:</p>

<pre><code class="language-bash">python mxfp8_sdpa_repro.py --baseline torch</code></pre>

<p class="p1">Specific sequence lengths with correctness checking:</p>

<pre><code class="language-bash">python mxfp8_sdpa_repro.py \
    --baseline torch \
    --seq-lens 8192 16384 \
    --heads-q 88 \
    --heads-kv 8 \
    --head-dim 128 \
    --check</code></pre>

<p class="p1">Padding-mask benchmark:</p>

<pre><code class="language-bash">python mxfp8_sdpa_repro.py \
    --baseline torch \
    --seq-lens 8192 \
    --skip-bwd \
    --padding-mask</code></pre>

<p class="p1">The script prints the environment used for each run so benchmark output is self-describing.</p>

<h2><span class="s1"><b>Questions</b></span></h2>

<ol start="1"><li>Is the current MXFP8 backward throughput on GB300 expected?</li><li>Are there planned optimizations for <code>sdpa_mxfp8_backward</code> on SM103?</li><li><span class="s1">Could </span><code>use_padding_mask</code><span class="s1"> / </span><code>seq_len_q</code><span class="s1"> / </span><code>seq_len_kv</code><span class="s1"> be exposed directly by </span><code>sdpa_mxfp8</code><span class="s1">?</span></li><li>Is the ~35% overhead from enabling the padding-mask path expected?</li><li>Are there recommended graph settings or tensor layouts that avoid this overhead today?</li></ol>

<h2><span class="s1"><b>Reproducer</b></span></h2>

<p class="p4">Attached:</p>

<pre><code class="language-text">mxfp8_sdpa_repro.py</code></pre>

<p class="p1">The reproducer is standalone and depends only on PyTorch and <code>nvidia-cudnn-frontend</code> when using the PyTorch BF16 baseline.</p>

</body>
</html>

[mxfp8_sdpa_repro.py](https://github.com/user-attachments/files/31203966/mxfp8_sdpa_repro.py)

**Comment by vedaanta** [comment]
Thanks @ZejiaZheng for the request! 😄 

MXFP8 bwd perf:
- Its not possible to optimize further easily without sacrificing numerics.
- We need to online scale some intermediate tensors, and that eats put all theoretical FP8 TC gains.
- If it works for your usecase, we can consider fixed scaling for dS, and that should provide more TFLops.

MXFP8 varlen:
- For fprop, lemme look into supporting it, as it helps more customers in general.
